### PR TITLE
Preserve forced open paths during loading of sfd

### LINKED ISSUE #3302: Preserve open paths during loading
Using FontForge 20170731 (Debian Testing/Buster).

The following python script demonstrates the issue.
```python
import fontforge

myfont1 = fontforge.font()

glyph = myfont1.createChar(65, "A")
pen = glyph.glyphPen()
pen.moveTo((20,20))
pen.lineTo((20,60))
pen.lineTo((60,60))
pen.closePath()

glyph = myfont1.createChar(66, "B")
pen = glyph.glyphPen()
pen.moveTo((20,20))
pen.lineTo((20,60))
pen.lineTo((60,60))
pen.lineTo((20,20))
pen.closePath()
pen.moveTo((100,100))
pen.lineTo((100,500))
pen.lineTo((500,500))
pen.lineTo((100,100))
pen.closePath()

glyph = myfont1.createChar(67, "C")
pen = glyph.glyphPen()
pen.moveTo((20,20))
pen.lineTo((20,60))
pen.lineTo((60,60))
pen.lineTo((20,20))
pen.endPath()
pen.moveTo((100,100))
pen.lineTo((100,500))
pen.lineTo((500,500))
pen.lineTo((100,100))
pen.endPath()

glyph = myfont1.createChar(68, "D")
pen = glyph.glyphPen()
pen.moveTo((20,20))
pen.lineTo((20,60))
pen.lineTo((60,60))
pen.endPath()
pen.moveTo((100,100))
pen.lineTo((100,500))
pen.lineTo((500,500))
pen.endPath()

glyph = myfont1.createChar(69, "E")
pen = glyph.glyphPen()
pen.moveTo((100,100))
pen.lineTo((100,500))
pen.lineTo((500,500))
pen.lineTo((100,100))
pen.endPath()

myfont1.fontname="TestFont1"
myfont1.save("TestFont1.sfd")
myfont1.generate("TestFont1.ttf")

myfont2 = fontforge.open("TestFont1.sfd") 

myfont2.save("TestFont2.sfd")
myfont2.generate("TestFont2.ttf")

myfont2.compareFonts(myfont1, "-", "outlines")
```
Since the second font is the same as the first one, the output should be empty. Instead I get:
```
Outline Glyphs
 Glyph Differences
  Open/Closed contour mismatch in glyph “C”
  Open/Closed contour mismatch in glyph “E”
```

P.S. @monkeyiq, thanks for introducing forced open path support (https://github.com/fontforge/fontforge/commit/3e7205354836e7d171b93619ff72b7cbc841eedf) in FontForge!

### PR DESCRIPTION
**Pull request by andreygursky**
If a contour is flagged with forced open path, it must not be auto closed.

Closes #3302

Provide a general summary of your changes in the **Title** above.

### Important
Mark with [x] to select. Leave as [ ] to unselect.

### Motivation and Context
- [ ] Why is this change required? What problem does it solve?
- [x] If it fixes an open issue, include the text `Closes #1` (where 1 would be the issue number) to your commit message.

### Types of changes
What types of changes does your code introduce? Check all the boxes that apply:
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Description
- [ ] Describe your changes in detail.

### Final checklist
Go over all the following points and check all the boxes that apply. 
If you're unsure about any of these, don't hesitate to ask. We're here to help! Various areas of the codebase have been worked on by different people in recent years, so if you are unfamiliar with the general area you're working in, please feel free to chat with people who have experience in that area. See the list [here](https://github.com/fontforge/fontforge/blob/master/CONTRIBUTING.md#people-to-ask).
- [ ] My code follows the code style of this project found [here](https://github.com/fontforge/fontforge/blob/master/CONTRIBUTING.md#coding-style).
- [x] My code follows the existing surrounding code style.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read some parts of the [**CONTRIBUTING**](https://github.com/fontforge/fontforge/blob/master/CONTRIBUTING.md) guidelines.

Your pull request will be tested via Travis CI to automatically indicate that your changes do not prevent compilation. FontForge is a big program, so Travis can easily take over 20 minutes to confirm your changes are buildable. Please be patient. More details about using Travis can be found [here](https://github.com/fontforge/fontforge/blob/master/CONTRIBUTING.md#using-travis-ci).  
  
If it reports back that there are problems, you can log into the Travis system and check the log report for your pull request to see what the problem was. If no error is shown, just re-run the Travis test for your pull-request (that failed) to see a fresh report since the last report may be for someone else that did a later pull request, or for mainline code. If you add new code to fix your issue/problem, then take note that you need to check the next pull request in the Travis system. Travis issue numbers are different from GitHub issue numbers.

### CODE REVIEW

**Comment by JoesCat** [comment]
Hi @andreygursky, I'm looking at this patch and thinking..., ...if someone were to be looking at code a couple months later, they might not see what's happening here, or the reasoning for some code. Would it be possible to sprinkle some comments to help other developers understand what they're looking at here? FontForge is a bit under-commented, so whatever you can add will certainly help.

Also, would you be interested in adding some test code that would pass with the fixes, and fail if someone accidentally reverted something? This would help safeguard what you've done here.
If you're not sure what I mean by tests, you can look at this [names2_patch](https://github.com/fontforge/fontforge/commit/1d9fb33853fa68bb70d72c5d3e2b88bcb4805163) - specifically around tests/testsuite.at, tests/Makefile.am, test1010.py

ping - @frank-trampe, @jtanx ?

**Comment by andreygursky** [comment]
Hi @JoesCat, yeah, in order to write a font converter in C I had to use fontforge/python.c as example for FontForge API, declare extern functions and the only help was the function names. I'm not sure I can do here anything.

I'll add my python reproducer to tests as perhaps fontforge script for (a bit) faster running, this would be useful, thanks for asking.

Would be not bad to have the example from https://github.com/fontforge/fontforge/issues/1264 as at least self-contained one. https://github.com/fontforge/fontforge/blob/9d425976e21e725113a359a53aa1a86b25829bb1/fontforge/sfd.c#L4052-L4058 looks like a no-op?

BTW, how can I download the work done by appveyor: fontforgebuilds\FontForge-mingw-w64-i686-9d4259-gdk2-appveyor.7z (and future rebuilds)? The log say it has been uploaded somewhere.

**Comment by frank-trampe** [comment]
@andreygursky, do you have a sample file handy that demonstrates the problem that you're trying to fix?

**Comment by andreygursky** [comment]
@frank-trampe, sure, I've started with the description in the issue https://github.com/fontforge/fontforge/issues/3302.

**Comment by andreygursky** [comment]
I've scrolled too fast to the end to see what happens there and overlooked these tabs at the top. @JoesCat, thanks!

File: fontforge/sfd.c [hunk 1]
```diff
@@ -3982,8 +3984,8 @@ static SplineSet *SFDGetSplineSet(FILE *sfd,int order2) {
 		    SplinePointList *spl = chunkalloc(sizeof(SplinePointList));
 		    spl->first = spl->last = pt;
 		    spl->start_offset = 0;
-		    if ( cur!=NULL ) {
-			if ( SFDCloseCheck(cur,order2))
+		    if ( cur!=NULL && !(have_cur_val==1 && cur_val & SFD_PTFLAG_FORCE_OPEN_PATH) ) {
+			if ( SFDCloseCheck(cur,order2) )
 			    --ttfindex;
```
Comment (frank-trampe) [inline fontforge/sfd.c]: If there is a "move" operation, we need to reset the path state since we are on a new path now, right?

Comment (frank-trampe) [reply fontforge/sfd.c]: @andreygursky, ping.

Comment (frank-trampe) [reply fontforge/sfd.c]: @andreygursky, ping.

**Comment by ctrlcctrlv** [comment]
I am closing this as @andreygursky has not replied to a query which is blocking the merge in eleven months. If he returns, this may still be merged; I'm not rejecting this, but simply acknowledging that without a reply from @andreygursky, who seems to have no intention of replying, this is unlikely to ever be merged.

**Pull request closed without merging**